### ISSUE TITLE
Request: Add support for custom vertex shaders

### ISSUE THREAD

**Issue by odedro987** [hajimehoshi/ebiten#2209]
Currently Ebitengine only supports fragments shaders. While fragments shaders give you a lot of freedom, there are some effects that are easier(and maybe more efficient) to implement using the vertex shader(scaling, stretching, rotating, changing the shape, etc.).
Rotating a handful of vertices is much simpler than all of the pixels within a source image.

This feature will give more more freedom, especially with combination with fragment shaders, allowing for more complex shaders to be possible.

Another more concrete example is this simple shader, moving the top vertices in a certain wave form to simulate wind - can be useful for moving grass sprites, trees, etc.
A possible implementation:
```go
var Amplitude float
var Frequency float
var Delay     float
var Time      float

func Vertex(position vec2, texCoord vec2, color vec4) (position vec4, texCoord vec2, color vec4) {
    // Check if current vertex is one of the top two vertices.
    // If using something similar to gl_VertexID is possible.
    // Else using texCoord it should also be possible to infer the vertex id.
    if texCoord.y == 0 { // Assuming (0,0) and (1, 0) correspond to the top vertices.
        position.x += Amplitude * sin((Time + Delay) * Frequency)
        //position.y += AmplitudeY * sin((Time + DelayY) * FrequencyY)
    }

	return __projectionMatrix * vec4(position, 0, 1), texCoord, color
}
```

One point raised, was that this is essentially exposing another API which then will need to be maintained, thus raising the maintenance cost. This is can be potentially flagged as experimental, and be left for the discretion of the user, whether to use a custom vertex shader or not.

**Comment by hajimehoshi** [comment]
Interesting, but can we already do similar effects by modifying vertices on the CPU side?

**Comment by SolarLune** [comment]
> Interesting, but can we already do similar effects by modifying vertices on the CPU side?

Yeah, that's true, we can.

In addition, scaling, stretching, and rotating are all possible with the current `GeoM` struct when drawing using images. I'm sure vertex shaders would be more efficient, as this would allow us to perform these calculations on the GPU and free up time on the CPU, but by how much I'm not sure.

I think it might be good to examine this from a performance perspective to see if, say, rotating, scaling, stretching, and skewing using GeoM is significantly worse than doing it through a theoretical vertex shader.

**Comment by hajimehoshi** [comment]
Yeah, we need to do some performance experiments. My concern is that a vertex shader might prevent batching, so the performance might be worse in some cases.

**Comment by Zyko0** [comment]
@SolarLune vertex shaders could allow bump mapping in Tetra?
Among many other things I guess, but I think this one is a common use case in 3D

**Comment by SolarLune** [comment]
> @SolarLune vertex shaders could allow bump mapping in Tetra? Among many other things I guess, but I think this one is a common use case in 3D

I suppose you could use a vertex shader to vary the shape of a mesh, though I think normal mapping would be most effectively done through modification of Tetra3D's rendering shader, rather than using vertex shaders for bump mapping (unless I am missing alternate possibilities).

Using a vertex shader would allow for rendering meshes faster in Tetra3D, but I would need to send the necessary data to transform each vertex to be able to take advantage of it (which Kage doesn't currently support, so that's not an option currently).

**Comment by svatez** [comment]
Will it be implemented?

**Comment by hajimehoshi** [comment]
Not planned

**Comment by sedyh** [comment]
I really want to see this feature in the engine, without it its impossible to make anything other than topdown games without dropping performance to zero.

**Comment by odedro987** [comment]
I actually kind of forgot about this issue haha. But yes, I would love to see this. You initially asked for a reason why we should add this and I could only give a basic example. After I dove into shaders a little bit more, I can give a better one. I recently did a shader for 3D rotations of a 2D sprite and as a result I had to correct the texture mapping. While this could be achieved with fragment shaders I believe it is much more compact passing the information through the vertex shader.

**Comment by SolarLune** [comment]
> I really want to see this feature in the engine, without it its impossible to make anything other than topdown games (isometry, voxels, spritestacking) without dropping performance to zero.

I'm in general support of custom vertex shaders and a better ability to run things on the GPU, but "impossible to make anything other than top-down games" is a heavy claim. Can you elaborate? What are you trying to make that would benefit heavily from vertex shaders?

**Comment by sedyh** [comment]
> Can you elaborate? What are you trying to make that would benefit heavily from vertex shaders?

Well, yes, not really impossible, but it can be quite slow. Recently, my colleagues and I were studying some public 2.5 engines and we began to understand that such engines do not scale well.

For example, here is what the author fauxtron 3d writes, who spent some time implementing a working engine for sprite stacking:
> Yeah, it's a 'fake 3d' engine but I use actual vertex buffers and projection matrices, so performance is actually maintainable. :) I plan to add in OBJ importation eventually so you can go the more 'true' 3d route if you wish. So this is actual 3D. The only sprite stack I have in this project is the fog, but that is also just a frozen vertex buffer (mesh) with a single surface as the texture for each layer (64 total layers). The buildings are all in a single buffer. Using standard 'draw_sprite()' is waayyy to slow to produce anything usable. haha. If you can batch things into a single sprite/buffer, that is the best way to get the most optimal performance. :)

**Comment by svatez** [comment]
Many engines have vertex shaders. I would like to port the code from my projects to Ebitengine without any using tricks on the CPU side.

**Comment by hajimehoshi** [comment]
@anthdm requested this to treat a lot of (about 5M?) vertices. I think this makes sense if you want to render such a number of vertices. I'll prioritize this issue.

**Comment by sedyh** [comment]
That's cool, is there a thread somewhere?

**Comment by hajimehoshi** [comment]
Unfortunately no, this was not an open discussion.

**Comment by sedyh** [comment]
Hmm, but they will be implemented, won't they?

**Comment by hajimehoshi** [comment]
There is no guarantee about this.
